File: core/src/main/java/com/emarsys/core/connection/ConnectionWatchDog.java
package com.emarsys.core.connection;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Handler;

import com.emarsys.core.util.Assert;

public class ConnectionWatchDog {

    private IntentFilter intentFilter = new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION);
    private BroadcastReceiver receiver;
    private Handler coreSdkHandler;

    ConnectivityManager connectivityManager;

    Context context;

    protected ConnectionWatchDog() {
    }

    public ConnectionWatchDog(Context context, Handler coreSdkHandler) {
        Assert.notNull(context, "Context must not be null!");
        Assert.notNull(coreSdkHandler, "CoreSdkHandler must not be null!");

        this.context = context.getApplicationContext();
        this.coreSdkHandler = coreSdkHandler;
        this.connectivityManager = (ConnectivityManager) this.context.getSystemService(Context.CONNECTIVITY_SERVICE);
    }

    public ConnectionState getConnectionState() {
        NetworkInfo activeNetwork = connectivityManager.getActiveNetworkInfo();

        if (activeNetwork == null || !activeNetwork.isConnected()) {
            return ConnectionState.DISCONNECTED;
        } else if (ConnectivityManager.TYPE_MOBILE == activeNetwork.getType()
                || ConnectivityManager.TYPE_MOBILE_DUN == activeNetwork.getType()) {
            return ConnectionState.CONNECTED_MOBILE_DATA;
        }

        return ConnectionState.CONNECTED;
    }

    public boolean isConnected() {
        NetworkInfo networkInfo = connectivityManager.getActiveNetworkInfo();
        return networkInfo != null && networkInfo.isConnected();
    }

    public void registerReceiver(ConnectionChangeListener connectionChangeListener) {
        if (receiver == null) {
            receiver = new ConnectivityChangeReceiver(connectionChangeListener);
            context.registerReceiver(receiver, intentFilter);
        } else {
            throw new IllegalStateException("Register receiver cannot be called multiple times!");
        }
    }

    class ConnectivityChangeReceiver extends BroadcastReceiver {

        ConnectionChangeListener connectionChangeListener;

        public ConnectivityChangeReceiver(ConnectionChangeListener connectionChangeListener) {
            Assert.notNull(connectionChangeListener, "ConnectionChangeListener must not be null!");

            this.connectionChangeListener = connectionChangeListener;
        }

        @Override
        public void onReceive(Context context, Intent intent) {
            final ConnectionState connectionState = getConnectionState();
            final boolean isConnected = isConnected();

            coreSdkHandler.post(new Runnable() {
                @Override
                public void run() {
                    connectionChangeListener.onConnectionChanged(connectionState, isConnected);
                }
            });
        }

    }

}

File: core/src/main/java/com/emarsys/core/connection/ConnectionWatchDog.kt
package com.emarsys.core.connection

import android.content.BroadcastReceiver
import android.content.Context
import android.content.IntentFilter
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import com.emarsys.core.Mockable
import com.emarsys.core.handler.ConcurrentHandlerHolder
import com.emarsys.core.util.AndroidVersionUtils

@Mockable
class ConnectionWatchDog(
    inputContext: Context,
    private val concurrentHandlerHolder: ConcurrentHandlerHolder
) : ConnectivityManager.NetworkCallback() {
    private companion object {
        var receiver: BroadcastReceiver? = null
    }

    private val context = inputContext.applicationContext
    private val intentFilter = IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION)
    private val connectivityManager =
        context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

    private val networkRequest: NetworkRequest =
        NetworkRequest.Builder()
            .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
            .addTransportType(NetworkCapabilities.TRANSPORT_ETHERNET)
            .addTransportType(NetworkCapabilities.TRANSPORT_VPN)
            .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
            .build()

    private lateinit var connectionChangeListener: ConnectionChangeListener

    val connectionState: ConnectionState
        get() {
            return try {
                val activeNetwork =
                    connectivityManager.activeNetwork ?: return ConnectionState.DISCONNECTED
                val networkCapabilities =
                    connectivityManager.getNetworkCapabilities(activeNetwork)
                        ?: return ConnectionState.DISCONNECTED
                when {
                    networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> ConnectionState.CONNECTED
                    networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN) -> ConnectionState.CONNECTED
                    networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> ConnectionState.CONNECTED
                    networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> ConnectionState.CONNECTED_MOBILE_DATA

                    else -> ConnectionState.DISCONNECTED
                }
            } catch (ignored: Exception) {
                ConnectionState.DISCONNECTED
            }
        }
    val isConnected: Boolean
        get() {
            return try {
                val network = connectivityManager.activeNetwork ?: return false
                val networkCapabilities =
                    connectivityManager.getNetworkCapabilities(network) ?: return false
                return when {
                    networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> true
                    networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> true
                    networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN) -> true
                    networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> true
                    else -> false
                }
            } catch (ignored: Exception) {
                false
            }
        }


    fun registerReceiver(connectionChangeListener: ConnectionChangeListener) {
        try {
            if (AndroidVersionUtils.isOreoOrAbove) {
                this.connectionChangeListener = connectionChangeListener
                connectivityManager.registerNetworkCallback(
                    networkRequest,
                    this,
                    concurrentHandlerHolder.coreHandler.handler
                )
            } else {
                if (receiver != null) {
                    receiver =
                        ConnectivityChangeReceiver(
                            connectionChangeListener,
                            this,
                            concurrentHandlerHolder
                        )
                    context.registerReceiver(receiver, intentFilter)
                }
            }
        } catch (ignored: Exception) {
        }
    }

    override fun onCapabilitiesChanged(network: Network, networkCapabilities: NetworkCapabilities) {
        super.onCapabilitiesChanged(network, networkCapabilities)
        val connectionState = connectionState
        val isConnected = isConnected
        connectionChangeListener.onConnectionChanged(connectionState, isConnected)
    }
}